File: rl/strategy/two_player_strategy.py
from typing import List, Optional

from engine import Strategy
from engine.card import Card, Suit
from engine.history import GameHistory
from engine.trick import Trick


class TwoPlayerStrategy(Strategy):
    def choose_card(
            self, hand: List[Card], game_history: GameHistory, trick: Trick, trump_suit: Suit
    ) -> int:
        is_first_player = len(trick.plays) == 0

        if is_first_player:
            return hand.index(self.play_highest_non_trump_card(hand, trump_suit))

        leading_card = trick.plays[0][1]
        card = (self.play_closest_higher_card(hand, leading_card)
                or self.play_lowest_card(hand, leading_card)
                or self.play_lowest_trump_card(hand, trump_suit)
                or self.play_lowest_card(hand))
        return hand.index(card)

    def play_highest_non_trump_card(self, hand: List[Card], trump_suit: Suit) -> Card:
        non_trump_cards = [card for card in hand if card.suit != trump_suit]
        if non_trump_cards:
            return max(non_trump_cards, key=lambda c: c.rank)
        return self.play_lowest_card(hand)

    def play_lowest_non_trump_card(self, hand: List[Card], trump_suit: Suit) -> Card:
        non_trump_cards = [card for card in hand if card.suit != trump_suit]
        if non_trump_cards:
            return min(non_trump_cards, key=lambda c: c.rank)
        return self.play_lowest_card(hand)

    def play_closest_higher_card(  self, hand: List[Card], leading_card: Card ) -> Optional[Card]:
        candidates = [card for card in hand if card.suit == leading_card.suit and card.rank > leading_card.rank]
        if candidates:
            return min(candidates, key=lambda c: c.rank)
        return None

    def play_lowest_card(self, hand: List[Card], leading_card: Optional[Card] = None) -> Card:
        if leading_card:
            same_suit_cards = [card for card in hand if card.suit == leading_card.suit]
            if same_suit_cards:
                return min(same_suit_cards, key=lambda c: c.rank)
        return min(hand, key=lambda c: c.rank)

    def play_lowest_trump_card(self, hand: List[Card], trump_suit: Suit) -> Optional[Card]:
        trump_cards = [card for card in hand if card.suit == trump_suit]
        if trump_cards:
            return min(trump_cards, key=lambda c: c.rank)
        return None

File: engine/src/main/java/com/mjaseem/trick/strategy/TwoPlayerStrategy.java
package com.mjaseem.trick.strategy;

import com.mjaseem.trick.engine.*;

import java.util.*;

public class TwoPlayerStrategy implements Strategy {

    private final int version;

    public TwoPlayerStrategy(int version) {
        this.version = version;
    }

    @Override
    public int chooseCard(List<Card> hand, GameHistory history, Trick trick, Records.Suit trumpSuit) {
        // Determine if we are the first player (no previous plays)
        boolean isFirstPlayer = trick.plays().isEmpty();

        if (isFirstPlayer) {
            if (version == 4) {
                return selectMidRangeNonTrumpCard(history, hand, trumpSuit).orElseGet(() -> playLowestCard(hand));
            } else if (version == 3) {
                return playHighestNonTrumpCard(hand, trumpSuit).orElseGet(() -> playLowestCard(hand));
            } else if (version == 2 && hand.size() < 8) {
                return playHighestNonTrumpCard(hand, trumpSuit).orElseGet(() -> playLowestCard(hand));
            } else {
                return playLowestNonTrumpCard(hand, trumpSuit).orElseGet(() -> playLowestCard(hand));
            }
        } else {
            // If it's not the first player, handle based on the leading card
            Card leadingCard = trick.plays().getFirst().getValue();

            // Play the closest higher card in the leading suit, or the lowest if none higher
            return playClosestHigherCard(hand, leadingCard)
                    .or(() -> playLowestCard(hand, leadingCard))
                    .or(() -> playLowestTrumpCard(hand, trumpSuit))
                    .orElseGet(() -> playLowestCard(hand));
        }
    }

    private Optional<Integer> selectMidRangeNonTrumpCard(GameHistory history, List<Card> hand, Records.Suit trumpSuit) {

        Records.Suit weakSuit = findWeakSuitFromHistory(history, trumpSuit);
        List<Card> weakSuitCards = filterCardsBySuit(hand, weakSuit);

        if (!weakSuitCards.isEmpty()) {
            // Play a high card from opponent's weak suit
            return Optional.of(hand.indexOf(Collections.max(weakSuitCards, Comparator.comparingInt(Card::rank))));
        }

        // Sort cards by rank
        List<Card> cards = filterCardsExcludingSuits(hand, Collections.singletonList(trumpSuit));
        if (cards.isEmpty()) return Optional.empty();
        Card card = cards.stream()
                .min(Comparator.comparingInt(c -> Math.abs(c.rank() - 7)))
                .orElseThrow(() -> new IllegalStateException("Hand should never be empty."));
        // Return the middle card
        return Optional.of(hand.indexOf(card));
    }

    private List<Card> filterCardsBySuit(List<Card> hand, Records.Suit suit) {
        if (suit == null) return Collections.emptyList();
        List<Card> filtered = new ArrayList<>();
        for (Card card : hand) {
            if (card.suit() == suit) {
                filtered.add(card);
            }
        }
        return filtered;
    }

    private List<Card> filterCardsExcludingSuits(List<Card> hand, List<Records.Suit> suits) {
        List<Card> filtered = new ArrayList<>();
        for (Card card : hand) {
            if (!suits.contains(card.suit())) {
                filtered.add(card);
            }
        }
        return filtered;
    }

    private Optional<Integer> playHighestNonTrumpCard(List<Card> hand, Records.Suit trumpSuit) {
        return hand.stream()
                .filter(card -> card.suit() != trumpSuit)
                .max(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf);
    }

    private Optional<Integer> playLowestNonTrumpCard(List<Card> hand, Records.Suit trumpSuit) {
        return hand.stream()
                .filter(card -> card.suit() != trumpSuit)
                .min(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf);
    }

    private Optional<Integer> playClosestHigherCard(List<Card> hand, Card leadingCard) {
        return hand.stream()
                .filter(card -> card.suit() == leadingCard.suit())
                .filter(card -> card.rank() > leadingCard.rank())
                .min(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf);
    }

    private Optional<Integer> playLowestCard(List<Card> hand, Card leadingCard) {
        return hand.stream()
                .filter(card -> card.suit() == leadingCard.suit())
                .min(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf);
    }


    private Optional<Integer> playLowestTrumpCard(List<Card> hand, Records.Suit trumpSuit) {
        return hand.stream()
                .filter(card -> card.suit() == trumpSuit)
                .min(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf);

    }

    private Records.Suit findWeakSuitFromHistory(GameHistory history, Records.Suit trumpSuit) {
        // Analyze history to find suits opponent is weak in
        Map<Records.Suit, Integer> opponentSuitCounts = countOpponentSuitPlays(history);
        return opponentSuitCounts.entrySet().stream()
                .min(Comparator.comparingInt(Map.Entry::getValue))
                .map(Map.Entry::getKey)
                .filter(x -> !x.equals(trumpSuit))
                .orElse(null);
    }

    private Map<Records.Suit, Integer> countOpponentSuitPlays(GameHistory history) {
        Map<Records.Suit, Integer> suitCounts = new HashMap<>();
        for (Trick trick : history.getTricks()) {
            for (Map.Entry<Player, Card> play : trick.plays()) {
                Records.Suit suit = play.getValue().suit();
                suitCounts.put(suit, suitCounts.getOrDefault(suit, 0) + 1);
            }
        }
        return suitCounts;
    }

    private int playLowestCard(List<Card> hand) {
        return hand.stream()
                .min(Comparator.comparingInt(Card::rank))
                .map(hand::indexOf)
                .orElseThrow(() -> new IllegalStateException("No card to play"));
    }
}
